lib/api/unittest/CDataFrameAnalyzerTrainingTest.cc

```diff
@@ -817,4 +817,128 @@ BOOST_AUTO_TEST_CASE(testCategoricalFieldsEmptyAsMissing) {
     });
 }
 
+BOOST_AUTO_TEST_CASE(testRunBoostedTreeRegressionFeatureImportance) {
+
+    // Test that feature importance correctly recognize the impact of regressors in a linear model.
+
+    double alpha{2.0};
+    double lambda{1.0};
+    double gamma{10.0};
+    double softTreeDepthLimit{5.0};
+    double softTreeDepthTolerance{0.1};
+    double eta{0.9};
+    std::size_t maximumNumberTrees{1};
+    double featureBagFraction{1.0};
+
+    int rows = 200;
+
+    // test with all  shap values
+    {
+        std::stringstream output;
+        auto outputWriterFactory = [&output]() {
+            return std::make_unique<core::CJsonOutputStreamWrapper>(output);
+        };
+
+        api::CDataFrameAnalyzer analyzer{
+            test::CDataFrameAnalysisSpecificationFactory::predictionSpec(
+                "regression", "c5", rows, 5, 4000000, 0, 0, {"c1"}, alpha,
+                lambda, gamma, softTreeDepthLimit, softTreeDepthTolerance, eta,
+                maximumNumberTrees, featureBagFraction, 4),
+            outputWriterFactory};
+
+        TDoubleVec expectedPredictions;
+
+        TStrVec fieldNames{"c1", "c2", "c3", "c4", "c5", ".", "."};
+        TStrVec fieldValues{"", "", "", "", "", "0", ""};
+        test::CRandomNumbers rng;
+        TDoubleVec weights{50, 150, 50, -50};
+
+        TDoubleVec values;
+        rng.generateUniformSamples(-10.0, 10.0, weights.size() * rows, values);
+
+        // make last column categorical
+        for (auto it = values.begin(); it < values.end(); it += 4) {
+            *it = (*it < 0) ? -10 : 10;
+        }
+
+        auto frame = setupLinearRegressionData(fieldNames, fieldValues,
+                                               analyzer, weights, values);
+        analyzer.handleRecord(fieldNames, {"", "", "", "", "", "", "$"});
+
+        rapidjson::Document results;
+        rapidjson::ParseResult ok(results.Parse(output.str()));
+        BOOST_TEST_REQUIRE(static_cast<bool>(ok) == true);
+
+        double c1_sum, c2_sum, c3_sum, c4_sum;
+        for (const auto& result : results.GetArray()) {
+            if (result.HasMember("row_results")) {
+                c1_sum += std::fabs(
+                    result["row_results"]["results"]["ml"]["shap.c1"].GetDouble());
+                c2_sum += std::fabs(
+                    result["row_results"]["results"]["ml"]["shap.c2"].GetDouble());
+                c3_sum += std::fabs(
+                    result["row_results"]["results"]["ml"]["shap.c3"].GetDouble());
+                c4_sum += std::fabs(
+                    result["row_results"]["results"]["ml"]["shap.c4"].GetDouble());
+            }
+        }
+        BOOST_TEST_REQUIRE(c2_sum > c1_sum);
+        BOOST_TEST_REQUIRE(c1_sum > c3_sum);
+        BOOST_TEST_REQUIRE(c1_sum > c4_sum);
+        BOOST_REQUIRE_CLOSE(c3_sum, c4_sum, 80); // c3 and c4 within 80% of each other
+    }
+
+    // test with 0 shap values
+    {
```

Comment:
Could we maybe factor out boilerplate code to set up the input data as a lambda?

Reply:
I created a fixture to deal with the boiler plate code